### PR TITLE
A support of a min(max) date.

### PR DESCRIPTION
**Pull request by DenBond7**
Added a support of a min(max) date. Added the MinMaxDateActivity to the sample project for a testing.

### CODE REVIEW

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 1]
```diff
@@ -7,16 +7,11 @@
 import android.os.Vibrator;
 import android.support.v4.app.DialogFragment;
 import android.text.format.DateUtils;
-import android.view.LayoutInflater;
-import android.view.View;
-import android.view.ViewGroup;
-import android.view.Window;
-import android.view.WindowManager;
+import android.view.*;
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:10]: This is a bad habit.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 2]
```diff
@@ -85,6 +81,9 @@
 
     private boolean mVibrate = true;
 
+  private Calendar minDateCalendar;
+  private Calendar maxDateCalendar;
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:85]: Why not keep with the style of the project and use the `m` prefix, as in `mMinDateCalendar`?

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 3]
```diff
@@ -97,7 +96,11 @@ public DatePickerDialog() {
         // Empty constructor required for dialog fragment. DO NOT REMOVE
     }
 
-	public static DatePickerDialog newInstance(OnDateSetListener onDateSetListener, int year, int month, int day) {
+  public static DatePickerDialog newInstance(OnDateSetListener onDateSetListener, Calendar calendar) {
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:99]: This breaks the API for existing users.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 4]
```diff
@@ -184,10 +187,9 @@ private void updateDisplay(boolean announce) {
 	}
 
 	private void updatePickers() {
-        Iterator<OnDateChangedListener> iterator = mListeners.iterator();
-        while (iterator.hasNext()) {
-            iterator.next().onDateChanged();
-        }
+    for (OnDateChangedListener mListener : mListeners) {
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:190]: This block seems unrelated to the min/max date feature.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 5]
```diff
@@ -195,12 +197,16 @@ public int getFirstDayOfWeek() {
 	}
 
 	public int getMaxYear() {
-		return mMaxYear;
-	}
+    if (maxDateCalendar == null) {
+      return mMaxYear;
+    } else return maxDateCalendar.get(Calendar.YEAR) >= getMinYear() ? maxDateCalendar.get(Calendar.YEAR) : getMinYear();
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:202]: Why not avoid this painful logic and remove the `minYear` and `maxYear` features? They are redundant now.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 6]
```diff
@@ -262,6 +268,9 @@ public View onCreateView(LayoutInflater layoutInflater, ViewGroup parent, Bundle
 			currentView = bundle.getInt(KEY_CURRENT_VIEW);
 			listPosition = bundle.getInt(KEY_LIST_POSITION);
 			listPositionOffset = bundle.getInt(KEY_LIST_POSITION_OFFSET);
+
+      minDateCalendar = (Calendar)bundle.getSerializable(KEY_MIN_DATE);
+      maxDateCalendar = (Calendar)bundle.getSerializable(KEY_MAX_DATE);
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:273]: Formatting is off.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 7]
```diff
@@ -330,6 +339,9 @@ public void onSaveInstanceState(Bundle bundle) {
 		bundle.putInt(KEY_YEAR_END, mMaxYear);
 		bundle.putInt(KEY_CURRENT_VIEW, mCurrentView);
 
+    bundle.putSerializable(KEY_MIN_DATE, minDateCalendar);
+    bundle.putSerializable(KEY_MAX_DATE, maxDateCalendar);
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:343]: Formatting is off.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java (hunk 1 above)
Comment (DenBond7) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:10]: Hello. I now. This is IDEA code optimizer. Sorry.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 8]
```diff
@@ -195,12 +197,16 @@ public int getFirstDayOfWeek() {
...
+      return mMaxYear;
+    } else return maxDateCalendar.get(Calendar.YEAR) >= getMinYear() ? maxDateCalendar.get(Calendar.YEAR) : getMinYear();
+  }
 
 	public int getMinYear() {
-		return mMinYear;
-	}
+    if (minDateCalendar == null) {
+      return mMinYear;
+    } else return minDateCalendar.get(Calendar.YEAR);
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:208]: More formatting issues.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java [hunk 9]
```diff
@@ -391,7 +403,48 @@ public void tryVibrate() {
 		}
 	}
 
-	static abstract interface OnDateChangedListener {
+  @Override
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DatePickerDialog.java:406]: This block is using 2-space indentation when the project uses 4-spaces.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DayPickerView.java [hunk 10]
```diff
@@ -125,6 +134,11 @@ public boolean goTo(SimpleMonthAdapter.CalendarDay day, boolean animate, boolean
 
 	public void init(Context paramContext) {
 		mContext = paramContext;
+
+    if (mController.getMinDate() != null){
+      mSelectedDay = mTempDay = new SimpleMonthAdapter.CalendarDay(mController.getMinDate().getTimeInMillis());
+    }
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DayPickerView.java:140]: More formatting issues.

File: datetimepicker-library/src/com/fourmob/datetimepicker/date/DayPickerView.java [hunk 11]
```diff
@@ -79,8 +81,15 @@ public boolean goTo(SimpleMonthAdapter.CalendarDay day, boolean animate, boolean
...
-        final int position = (day.year - mController.getMinYear())
+        int position = (day.year - mController.getMinYear())
                 * SimpleMonthAdapter.MONTHS_IN_YEAR + day.month;
+        if(mController.getMinDate() != null){
+          Calendar startCalendar = mController.getMinDate();
+          Calendar endCalendar = day.getCalendar();
+
+          int diffYear = endCalendar.get(Calendar.YEAR) - startCalendar.get(Calendar.YEAR);
+          position = diffYear * SimpleMonthAdapter.MONTHS_IN_YEAR + endCalendar.get(Calendar.MONTH) - startCalendar.get(Calendar.MONTH);
+        }
```
Comment (tadfisher) [inline datetimepicker-library/src/com/fourmob/datetimepicker/date/DayPickerView.java:92]: 2-space indents should be 4.
